### PR TITLE
Refactor roundrobin to support custom picker

### LINKED ISSUE #1690: reuse roundrobin balancer with custom picker?
I need to implement a custom balancer that uses rendezvous hashing to select the server (the key is passed in via the context)

I implemented this in the old balancer API by copying the round robin implementation. I would like to avoid this in the new API.

It seems like all I should have to do is implement a new picker. Does it make sense to refactor the roundrobin balancer so a picker can be provided?

### PR DESCRIPTION
**Pull request by menghanl**
fixes #1690

### CODE REVIEW

File: balancer/base/balancer.go [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+		csEvltr:  &connectivityStateEvaluator{},
+		// Initialize picker to a picker that always return
+		// ErrNoSubConnAvailable, because when state of a SubConn changes, we
+		// may call UpdateBalancerState with this picker.
+		picker: &errPicker{err: balancer.ErrNoSubConnAvailable},
+	}
+}
+
+func (*baseBuilder) Name() string {
+	return "round_robin"
```
Comment (predmond) [inline balancer/base/balancer.go]: I think the pickerBuilder needs to provide the name now.

Comment (menghanl) [reply balancer/base/balancer.go]: Thanks for taking a look. Fixed.

File: balancer/base/base.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package base defines a balancer base that can be used to build balancers with
+// different picking algothrim.
```
Comment (dfawley) [inline balancer/base/base.go]: algothrim -> algorithms

File: balancer/base/base.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+//
+// The base balancer create a new SubConn for each resolved address. The
```
Comment (dfawley) [inline balancer/base/base.go]: *creates

File: balancer/base/base.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+// provided picker will only be notified about READY SubConns.
+//
+// This package is the base of round_robin balancer, its purpose is to be used
+// to build round_robin like balancers with complex picking algothrim. Balancers
```
Comment (dfawley) [inline balancer/base/base.go]: *algorithms

File: balancer/base/balancer.go [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+	b.cc.UpdateBalancerState(b.state, b.picker)
+	return
+}
+
+// Close is a nop because base balancer doesn't internal state to clean
+// up, and it doesn't need to call RemoveSubConn for the SubConns.
+func (b *baseBalancer) Close() {
+}
+
+type errPicker struct {
```
Comment (dfawley) [inline balancer/base/balancer.go]: Even though it's trivial, we may want to export this for our users, too.

File: balancer/base/base.go [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+// with more complicated logic should try to implement a balancer builder from
+// scratch.
+package base
```
Comment (dfawley) [inline balancer/base/base.go]: EXPERIMENTAL for now?

File: balancer/base/balancer.go (hunk 5 above)
Comment (menghanl) [reply balancer/base/balancer.go]: Done

File: balancer/base/base.go (hunk 2 above)
Comment (menghanl) [reply balancer/base/base.go]: Done * 2

File: balancer/base/base.go (hunk 3 above)
Comment (menghanl) [reply balancer/base/base.go]: Done

File: balancer/base/base.go (hunk 6 above)
Comment (menghanl) [reply balancer/base/base.go]: Done

**Pull request merged**